tests/e2e/e2e_test.go

```diff
@@ -0,0 +1,268 @@
+	s.Require().NoError(err)
+
+	balanceTest1, err := node.QuerySpecificBalance(test1Addr, "uluna")
+	s.Require().NoError(err)
+
+	s.Require().Equal(balanceTest1.Amount, transferAmount1)
+	s.Require().Equal(newValidatorBalance, decremented)
+
+	// Test 2: try bank send with grant
+	test2Addr := node.CreateWallet("test2")
+	transferAmount2 := sdkmath.NewInt(10000000)
+	transferCoin2 := sdk.NewCoin("uluna", transferAmount2)
+
+	node.BankSend(transferCoin2.String(), validatorAddr, test2Addr)
+	node.GrantAddress(test2Addr, test1Addr, transferCoin2.String(), "test2")
+
+	validatorBalance, err = node.QuerySpecificBalance(validatorAddr, "uluna")
+	s.Require().NoError(err)
+
+	node.BankSendFeeGrantWithWallet(transferCoin2.String(), test1Addr, validatorAddr, test2Addr, "test1")
+
+	newValidatorBalance, err = node.QuerySpecificBalance(validatorAddr, "uluna")
+	s.Require().NoError(err)
+
+	balanceTest1, err = node.QuerySpecificBalance(test1Addr, "uluna")
+	s.Require().NoError(err)
+
+	balanceTest2, err := node.QuerySpecificBalance(test2Addr, "uluna")
+	s.Require().NoError(err)
+
+	s.Require().Equal(balanceTest1.Amount, transferAmount1.Sub(transferAmount2))
+	s.Require().Equal(newValidatorBalance, validatorBalance.Add(transferCoin2))
+	s.Require().Equal(balanceTest2.Amount, sdk.NewDecWithPrec(98, 2).MulInt(transferAmount2).TruncateInt())
+
+	// Test 3: banktypes.MsgMultiSend
+	validatorBalance, err = node.QuerySpecificBalance(validatorAddr, "uluna")
+	s.Require().NoError(err)
+
+	node.BankMultiSend(transferCoin1.String(), false, validatorAddr, test1Addr, test2Addr)
+
+	newValidatorBalance, err = node.QuerySpecificBalance(validatorAddr, "uluna")
+	s.Require().NoError(err)
+
+	subAmount = sdk.NewDecWithPrec(204, 2).MulInt(transferAmount1).TruncateInt()
+	s.Require().Equal(newValidatorBalance, validatorBalance.Sub(sdk.NewCoin("uluna", subAmount)))
+}
+
+func (s *IntegrationTestSuite) TestFeeTaxWasm() {
+	chain := s.configurer.GetChainConfig(0)
+	node, err := chain.GetDefaultNode()
+	s.Require().NoError(err)
+
+	testAddr := node.CreateWallet("test")
+	transferAmount := sdkmath.NewInt(100000000)
+	transferCoin := sdk.NewCoin("uluna", transferAmount)
+	node.BankSend(fmt.Sprintf("%suluna", transferAmount.Mul(sdk.NewInt(4))), initialization.ValidatorWalletName, testAddr)
+
+	node.StoreWasmCode("counter.wasm", initialization.ValidatorWalletName)
+	chain.LatestCodeID = int(node.QueryLatestWasmCodeID())
+	node.InstantiateWasmContract(
+		strconv.Itoa(chain.LatestCodeID),
+		`{"count": "0"}`, transferCoin.String(),
+		"test")
+
+	contracts, err := node.QueryContractsFromID(chain.LatestCodeID)
+	s.Require().NoError(err)
+	s.Require().Len(contracts, 1, "Wrong number of contracts for the counter")
+
+	balance, err := node.QuerySpecificBalance(testAddr, "uluna")
+	s.Require().NoError(err)
+	s.Require().Equal(balance.Amount, sdk.NewDecWithPrec(98, 2).MulInt(transferAmount).Add(sdk.NewDecFromInt(transferAmount.MulRaw(2))).TruncateInt())
```

Comment:
can you help me write down the reasoning for this calculation? I assume that this is for burn tax calculation?

something like this should be helpful

<img width="591" alt="Screenshot 2024-02-01 at 01 32 18" src="https://github.com/classic-terra/core/assets/23058164/ea255b43-ae37-41ed-9827-644169e1d9cb">